float8_experimental/dynamic_linear/dynamic_float8_linear.py

```diff
@@ -38,6 +39,20 @@ def backward(ctx, gradY):
             None,
         )
 
+def cast_x_to_float8_e4m3fn_pre_hook(module, args):
+    """
+    Hook to cast the incoming activation to `torch.float8_e4m3fn`
+    """
+    return module.cast_to_float8(args[0])
+
+
+def cast_dldy_to_float8_e5m2_forward_hook(module, args, output):
+    """
+    Hook to cast the incoming gradient to `torch.float8_e5m2`
+    """
+    new_output = NoopFwToFloat8E5M2Bw.apply(output, module.emulate)
```

Comment:
I'm a bit surprised that torch.compile does not support this case with autograd function, as both DTensor pre-forward hook and forward hook are using autograd functions, probably because DTensor is using sth like `allow_in_graph` in dynamo. Maybe one way to workaround is to use a function that calls `NoopFwToFloat8E5M2Bw.apply(output, module.emulate)` inside, and then make it a allow_in_graph in dynamo?